File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/proxy/Pursuit.java
package jv.com.tencent.cain.design_pattern.proxy;

/**
 * @author cainjiang
 * @date 2019-05-17
 * RealSubject
 */
public class Pursuit implements IGiveGift {
    private SchoolGirl schoolGirl;

    public Pursuit(SchoolGirl schoolGirl) {
        this.schoolGirl = schoolGirl;
    }

    public Pursuit() {
    }

    @Override
    public void giveDolls() {
        String name = "";
        if (schoolGirl != null) {
            name = schoolGirl.getName();
        }
        System.out.println(name + "送你洋娃娃🤩");
    }

    @Override
    public void giveFlowers() {
        String name = "";
        if (schoolGirl != null) {
            name = schoolGirl.getName();
        }
        System.out.println(name + "送你鲜花💐");
    }

    @Override
    public void giveChocolate() {
        String name = "";
        if (schoolGirl != null) {
            name = schoolGirl.getName();
        }
        System.out.println(name + "巧克力🍫");

    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/proxy/Pursuit.kt
package kt.com.tencent.cain.desigin_pattern.proxy

/**
 * @author : jiangyu
 * @date   : 2019-05-18
 * @desc   : 代理设计模式中的RealSubject
 */
class Pursuit(private val schoolGirl: SchoolGirl = SchoolGirl()) : IGiveGift {

    override fun giveDolls() {
        println("${schoolGirl.name}送你洋娃娃")
    }

    override fun giveFlowers() {
        println("${schoolGirl.name}送你鲜花💐")
    }

    override fun giveChocolate() {
        println("${schoolGirl.name}送你巧克力🍫")
    }
}